### DISCUSSION TITLE
Useful fzf preview

### DISCUSSION THREAD

**Discussion by S1M0N38** [papis/papis#511] [Show and tell]
# Proof of Concept
I think that fzf window preview can be used to display useful information about the selected document. Useful information can be directly extract from the corresponding `info.yaml` using programs such as [yq](https://github.com/mikefarah/yq) and then nicely formatted into the preview window. 

For example I use this idea to dispaly papers abstract in the preview window
Here are my fzf settings for papis:
```ini
[settings]
picktool = fzf
fzf-header-format = {c.Fore.GREEN}{doc[author_list][0][family]:11.11}{c.Style.RESET_ALL} {c.Fore.YELLOW}{doc[year]:4}{c.Style.RESET_ALL} {c.Fore.BLUE}{doc[title]}{c.Style.RESET_ALL}
fzf-extra-flags = ["--ansi", "--multi", "-i", "--preview", "yq '.abstract' $(papis list --info {})", "--preview-window", "right:wrap", "--color"]
```
![working-preview](https://user-images.githubusercontent.com/22257750/220608404-35cf6b98-b538-4ce7-83eb-f94a3f1ec136.png)

# Bug
I have very basic knowledge about bash and I wasn't able to properly escape special chars in the placeholder `{}` for the window preview command (i.e. `yq '.abstract' $(papis list --info {})`). So at the moment papers with `:` in the title raise an error
![not-working-preview](https://user-images.githubusercontent.com/22257750/220609567-e427ea38-f21f-4c7d-9334-9fc271985cdc.png)

# Future Improvements
I think that this concept of showing additional information in the preview window can be further improved. For example papers information parsed from `info.yaml` with `yq` can be formatted in a pretty and colorful way (e.g. title and year, new line , list of authors, new line, abstract).

> If you know how to fix the aforementioned bug let me know :)

**Comment by jghauser** [comment]
Hey! This is very interesting. I used to use something similar (though leveraging `jinja2` rather than `yq`). I vaguely remember at some point using a [`config.py`](https://papis.readthedocs.io/en/latest/configuration.html#python-configuration-file) configuration file because it was easier to handle escaping correctly. Maybe this might make things easier for you too!

The general structure of config.py is like this:
```python
import papis.config

options = {
    "settings": {
        "document-description-format": '{{doc["title"]}} - {{doc["author"]}}',
    }
}
papis.config.register_default_settings(options)
```

**Comment by S1M0N38** [comment]
Here is a solution:
I create a separate script (`preview.sh`) placed in `$XDG_CONFIG_HOME/papis/scripts/preview.sh`
```bash
#!/usr/bin/env bash

YAML=$(papis list --info "${1//:/}")

SUMMARY="\033[33m$(yq '.year' $YAML)\033[0m "
SUMMARY+="\033[34m$(yq '.title' $YAML)\033[0m\n\n"
SUMMARY+="\033[32m$(yq '.author' $YAML)\033[0m\n\n"
SUMMARY+="$(yq '.abstract' $YAML)"

echo -e "$SUMMARY"
```

which can be called by the following fzf-extra-flags
```ini
[settings]
picktool = fzf
fzf-header-format = {c.Fore.GREEN}{doc[author_list][0][family]:11.11}{c.Style.RESET_ALL} {c.Fore.YELLOW}{doc[year]:4}{c.Style.RESET_ALL} {c.Fore.BLUE}{doc[title]}{c.Style.RESET_ALL}
fzf-extra-flags = ["--ansi", "--multi", "-i", "--preview", "$XDG_CONFIG_HOME/papis/scripts/preview.sh {}", "--preview-window", "right:wrap"]
```

![Screenshot 2023-03-05 at 13 46 01](https://user-images.githubusercontent.com/22257750/222961516-084b627a-726c-4db6-9eb5-371f7af63683.png)
